Sources/PackageSyntax/PackageEditor.swift

```diff
@@ -0,0 +1,427 @@
+
+        if loadedManifest.targets.contains(where: { $0.name == newTarget.name }) {
+            context.diagnosticsEngine.emit(.targetAlreadyExists(name: newTarget.name,
+                                                                packageName: loadedManifest.name))
+            throw Diagnostics.fatalError
+        }
+
+        let manifestContents = try fs.readFileContents(manifestPath).cString
+        let editor = try ManifestRewriter(manifestContents, diagnosticsEngine: context.diagnosticsEngine)
+
+        switch newTarget {
+        case .library(name: let name, includeTestTarget: _, dependencyNames: let dependencyNames),
+             .executable(name: let name, dependencyNames: let dependencyNames),
+             .test(name: let name, dependencyNames: let dependencyNames):
+            try editor.addTarget(targetName: newTarget.name,
+                                 factoryMethodName: newTarget.factoryMethodName(for: loadedManifest.toolsVersion))
+
+            for dependency in dependencyNames {
+                if loadedManifest.targets.map(\.name).contains(dependency) {
+                    try editor.addByNameTargetDependency(target: name, dependency: dependency)
+                } else if let productPackage = productPackageNameMapping[dependency] {
+                    if productPackage == dependency {
+                        try editor.addByNameTargetDependency(target: name, dependency: dependency)
+                    } else {
+                        try editor.addProductTargetDependency(target: name, product: dependency, package: productPackage)
+                    }
+                } else {
+                    context.diagnosticsEngine.emit(.missingProductOrTarget(name: dependency))
+                    throw Diagnostics.fatalError
+                }
+            }
+        case .binary(name: let name, urlOrPath: let urlOrPath, checksum: let checksum):
+            guard loadedManifest.toolsVersion >= .v5_3 else {
+                context.diagnosticsEngine.emit(.unsupportedToolsVersionForBinaryTargets)
+                throw Diagnostics.fatalError
+            }
+            try editor.addBinaryTarget(targetName: name, urlOrPath: urlOrPath, checksum: checksum)
+        }
+
+        try context.verifyEditedManifest(contents: editor.editedManifest)
+        try fs.writeFileContents(manifestPath, bytes: ByteString(encodingAsUTF8: editor.editedManifest))
+
+        // Write template files.
+        try writeTemplateFilesForTarget(newTarget)
+
+        if case .library(name: let name, includeTestTarget: true, dependencyNames: _) = newTarget {
+            try self.addTarget(.test(name: "\(name)Tests", dependencyNames: [name]),
+                               productPackageNameMapping: productPackageNameMapping)
+        }
+    }
+
+    private func diagnoseUnsupportedToolsVersions(manifest: Manifest) throws {
+        guard manifest.toolsVersion >= .v5_2 else {
```

Comment:
Regarding earlier comment about versioned manifests:  I guess this error check would take care of the case of an old `Package.swift` with a new `Package-5.2.swift` etc, so might not need an additional check.